rasa/core/actions/action.py

```diff
@@ -636,27 +637,48 @@ def __init__(self, name: Text, action_endpoint: Optional[EndpointConfig]) -> Non
         self.action_endpoint = action_endpoint
 
     def _action_call_format(
-        self, tracker: "DialogueStateTracker", domain: "Domain"
+        self,
+        tracker: "DialogueStateTracker",
+        domain: "Domain",
     ) -> Dict[Text, Any]:
         """Create the request json send to the action server."""
         from rasa.shared.core.trackers import EventVerbosity
 
         tracker_state = tracker.current_state(EventVerbosity.ALL)
 
-        return {
+        result = {
             "next_action": self._name,
             "sender_id": tracker.sender_id,
             "tracker": tracker_state,
-            "domain": domain.as_dict(),
             "version": rasa.__version__,
         }
 
+        if (
+            not self._is_selective_domain_enabled()
+            or domain.does_action_explicitly_need_domain(self.name())
+        ):
+            result["domain"] = domain.as_dict()
+
+        return result
+
+    def _is_selective_domain_enabled(self) -> bool:
+        is_selective_domain_enabled = False
+        if self.action_endpoint is not None:
+            is_selective_domain_enabled = bool(
+                self.action_endpoint.kwargs.get(
+                    SELECTIVE_DOMAIN, DEFAULT_SELECTIVE_DOMAIN
+                )
+            )
+
+        return is_selective_domain_enabled
```

Comment:
I'd refactor this too, to simplify overindentation (probably needs some extra formatting again):
```suggestion
        if self.action_endpoint is None:
            # add a logger.debug to warn of this issue
            return False
        return bool(
                self.action_endpoint.kwargs.get(
                    SELECTIVE_DOMAIN, DEFAULT_SELECTIVE_DOMAIN
                )
            )
```